SteamKit2/SteamKit2/Steam/Discovery/SmartCMServerList.cs

```diff
@@ -91,45 +96,98 @@ private bool WaitForServersFetched()
+            }
+            else
+            {
+                DebugWrite( "Resolving server list using the provider" );
+            }
 
             IEnumerable<ServerRecord> serverList = await configuration.ServerListProvider.FetchServerListAsync().ConfigureAwait( false );
             IReadOnlyCollection<ServerRecord> endpointList = serverList.ToList();
 
-            if ( endpointList.Count == 0 && configuration.AllowDirectoryFetch )
+            // Provider server list is fresh enough and it provided servers
+            if ( endpointList.Count > 0 )
             {
-                DebugWrite( "Server list provider had no entries, will query SteamDirectory" );
-                endpointList = await SteamDirectory.LoadAsync( configuration ).ConfigureAwait( false );
+                DebugWrite( $"Resolved {endpointList.Count} servers from the provider" );
+                ReplaceList( endpointList, writeProvider: false, providerRefreshTime );
+                return;
             }
 
-            if ( endpointList.Count == 0 && configuration.AllowDirectoryFetch )
+            // If API fetch is not allowed, bail out with no servers
+            if ( !configuration.AllowDirectoryFetch )
             {
-                DebugWrite( "Could not query SteamDirectory, falling back to cm0" );
-                var cm0 = await Dns.GetHostAddressesAsync( "cm0.steampowered.com" ).ConfigureAwait( false );
+                DebugWrite( $"Server list provider had no entries, and {nameof( SteamConfiguration.AllowDirectoryFetch )} is false" );
+                ReplaceList( [], writeProvider: false, DateTime.MinValue );
+                return;
+            }
 
-                endpointList = cm0.Select( ipaddr => ServerRecord.CreateSocketServer( new IPEndPoint(ipaddr, 27017) ) ).ToList();
+            // If the force refresh tried to fetch the server list already, do not fetch it again
+            if ( !alreadyTriedDirectoryFetch )
+            {
+                DebugWrite( $"Server list provider had no entries, will query {nameof( SteamDirectory )}" );
+                endpointList = await SteamDirectory.LoadAsync( configuration ).ConfigureAwait( false );
+
+                if ( endpointList.Count > 0 )
+                {
+                    DebugWrite( $"Resolved {endpointList.Count} servers from {nameof( SteamDirectory )}" );
+                    ReplaceList( endpointList, writeProvider: true, DateTime.UtcNow );
+                    return;
+                }
             }
 
-            DebugWrite( "Resolved {0} servers", endpointList.Count );
-            ReplaceList( endpointList );
+            // This is a last effort to attempt any valid connection to Steam,
+            // however Steam client does not appear to do this anymore,
+            // and this will not work when Steam moves to only websocket servers
+            DebugWrite( $"Server list provider had no entries, {nameof( SteamDirectory )} failed, falling back to cm0" );
```

Comment:
I didn't check if this server still works, but I'm thinking this should be removed at some point.

Reply:
I'd consider hardcoding some general servers that *seem* like their name couldn't change.

It's better than nothing, even if the list gets outdated. It'd be fallback used only if fetching up-to-date info has failed. I'd rather thave 5 hardcoded servers than zero in such case. You could return them at random, never save, never do anything else.

Reply:
I added a public static string for default servers (had to make separate ones for WS / tcp)

Reply:
And now after last Steam update they changed the servers:

cmp1-dfw1.steamserver.net:27018
cmp1-dfw1.steamserver.net:443
cmp1-iad1.steamserver.net:27018
cmp1-iad1.steamserver.net:443
cmp1-lax1.steamserver.net:27018
cmp1-lax1.steamserver.net:443
cmp1-sea1.steamserver.net:27018
cmp1-sea1.steamserver.net:443